python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+            self._entity_fields
+        )
+
+    def canFetchMore(self, index):
+        """
+        Returns True if there is more data available for parent; otherwise
+        returns False.
+
+        :param index: The index of the item being tested.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+        """
+
+        if not index.isValid():
+            return False
+
+        # get the item and it's stored hierarchy data
+        item = self.itemFromIndex(index)
+
+        if item.data(self.SG_ITEM_FETCHED_MORE):
+            # more data has already been queried for this item
+            return False
+
+        item_data = get_sg_data(item)
+
+        if not item_data:
+            return False
+
+        # the number of existing child items
+        child_item_count = item.rowCount()
+
+        # we can fetch more if there are no children already and the item
+        # has children.
+        return child_item_count == 0 and item_data.get("has_children", False)
+
+    ############################################################################
+    # protected methods
+
+    def _get_queried_urls_r(self, item):
+        """
+        Returns a list of previously queried urls for items under the supplied
+        parent item.
+
+        This method is used to identify which items need to be re-queried when a
+        model refresh is requested. It recursively iterates over the supplied
+        item's children, looking for non-empty leaf nodes that have children.
+
+        :param item: The parent :class:`~PySide.QtGui.QStandardItem` instance
+            to use when checking for queried children.
+
+        :return: A list of urls
+        :rtype: list
+        """
+
+        urls = []
+
+        # iterate over all the child items
+        for row in range(item.rowCount()):
+
+            child_item = item.child(row)
+
+            data = get_sg_data(child_item)
+            if data.get("ref", {}).get("kind") == "empty":
+                # empty items don't have valid urls to query
+                continue
+
+            if child_item.rowCount() == 0:
+                # this is an existing leaf node whose (potential) children have
+                # not been queried. do nothing.
+                continue
+
+            urls.append(data["url"])
+
+            # recurse
+            urls.extend(self._get_queried_urls_r(child_item))
+
+        return urls
+
+    def _insert_subtree(self, nav_data):
+        """
+        Inserts a subtree for the item represented by ``nav_data``.
+
+        The method first creates the item, then attempts to update/populate its
+        children.
+
+        :param dict nav_data: A dictionary of item data as returned via async
+            call to ``nav_expand``.
+        """
+
+        item = self.__create_item(nav_data)
+        self._update_subtree(item, nav_data)
+
+    def _item_created(self, item):
+        """
+        Checks to see if the item is an "empty" item and sets the foreground
```

Comment:
This doc string looks a little strange to me - from a user's doc point of view, i think is quite confusing. Base class docs here:

https://github.com/shotgunsoftware/tk-framework-shotgunutils/pull/49/files#diff-d442a3297316abbae14a00adf3457696R504

I think it should at least refer to those docs somehow? Sometimes giving feedback around this sort of thing ie easier once you see the built sphinx docs.